### PR TITLE
Move some inline definitions out

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Inline definitions introduce dependencies on headers and slow down compilation.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/common/buffer.h [hunk 1]
```diff
@@ -51,36 +45,20 @@ class LocalMemoryBuffer : public Buffer {
...
   /// Construct a LocalMemoryBuffer of all zeros of the given size.
-  LocalMemoryBuffer(size_t size) : has_data_copy_(true) {
-    buffer_.resize(size, 0);
-    data_ = buffer_.data();
-    size_ = buffer_.size();
-  }
+  LocalMemoryBuffer(size_t size);
 
-  uint8_t *Data() const override { return data_; }
+  inline uint8_t *Data() const override { return data_; }
```
Comment (pcmoritz) [inline src/ray/common/buffer.h]: Let's remove the "inline" keyword from the methods that are already inlined (it only adds clutter).

File: src/ray/common/function_descriptor.h [hunk 2]
```diff
@@ -13,18 +13,15 @@ using FunctionDescriptorType = rpc::FunctionDescriptor::FunctionDescriptorCase;
 class FunctionDescriptorInterface : public MessageWrapper<rpc::FunctionDescriptor> {
  public:
   /// Construct an empty FunctionDescriptor.
-  FunctionDescriptorInterface() : MessageWrapper() {}
+  FunctionDescriptorInterface();
```
Comment (pcmoritz) [inline src/ray/common/function_descriptor.h]: All of these have to be declared inline, right?

File: src/ray/common/ray_config.cc [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#include "ray/common/ray_config.h"
```
Comment (pcmoritz) [inline src/ray/common/ray_config.cc]: Let's remove this file and put it back inline (I don't think there is a point of having this file).

File: src/ray/gcs/accessor.h [hunk 4]
```diff
@@ -156,7 +156,7 @@ class JobInfoAccessor {
 /// task information in the GCS.
 class TaskInfoAccessor {
  public:
-  virtual ~TaskInfoAccessor() {}
+  inline virtual ~TaskInfoAccessor() {}
```
Comment (pcmoritz) [inline src/ray/gcs/accessor.h]: combining inline with virtual makes no sense (virtual functions can't be inlined)

**Comment by mehrdadn** [comment]
Microbenchmark results:

Before:

	single client get calls per second 16719.46 +- 108.53
	single client put calls per second 6708.09 +- 62.11
	single client put gigabytes per second 1.19 +- 0.0
	multi client put calls per second 8179.68 +- 30.3
	multi client put gigabytes per second 5.0 +- 3.44
	single client tasks sync per second 2627.04 +- 11.16
	single client tasks async per second 18196.31 +- 71.32
	multi client tasks async per second 31780.89 +- 308.71
	1:1 actor calls sync per second 4361.31 +- 15.44
	1:1 actor calls async per second 11287.53 +- 80.01
	1:1 actor calls concurrent per second 10746.0 +- 189.99
	1:n actor calls async per second 18452.71 +- 357.25
	n:n actor calls async per second 39831.87 +- 92.05
	n:n actor calls with arg async per second 15123.17 +- 248.67

After:

	single client get calls per second 16667.55 +- 174.13
	single client put calls per second 6597.48 +- 63.57
	single client put gigabytes per second 1.16 +- 0.0
	multi client put calls per second 7984.79 +- 73.67
	multi client put gigabytes per second 4.85 +- 3.29
	single client tasks sync per second 2546.74 +- 30.81
	single client tasks async per second 17774.78 +- 26.32
	multi client tasks async per second 31018.66 +- 353.75
	1:1 actor calls sync per second 4229.99 +- 26.78
	1:1 actor calls async per second 11013.39 +- 172.73
	1:1 actor calls concurrent per second 10982.3 +- 91.72
	1:n actor calls async per second 17803.66 +- 311.05
	n:n actor calls async per second 38629.5 +- 399.96
	n:n actor calls with arg async per second 14806.9 +- 167.62

Comparison:

	Before: 	After:  	Delta:
	16719.46	16667.55	-0.3%
	 6708.09	 6597.48	-1.6%
	    1.19	    1.16	-2.5%
	 8179.68	 7984.79	-2.4%
	     5.0	    4.85	-3.0%
	 2627.04	 2546.74	-3.1%
	18196.31	17774.78	-2.3%
	31780.89	31018.66	-2.4%
	 4361.31	 4229.99	-3.0%
	11287.53	11013.39	-2.4%
	 10746.0	 10982.3	 2.2%
	18452.71	17803.66	-3.5%
	39831.87	 38629.5	-3.0%
	15123.17	 14806.9	-2.1%

**Comment by ericl** [comment]
This is a pretty big performance impact. Can you only move out large chunks of code or those that read guarded variables? There is a reason we prefer inlining certain things.

Also, this PR needs to be split up into smaller pieces, as is it is quite large and will be infeasible to revert should it be necessary.

**Comment by ericl** [comment]
I would probably break it into a few phases like this:
1. Within each sub-package, move things from .h to .cc (without needing to change BUILD files). Avoid crossing package boundaries in each of these PRs so it is easier to isolate / revert regression (microbenchmarks would be helpful too).
2. Then, PRs that involve BUILD changes.

That would be easier to review and have less potential risk to maintenance activities.

**Comment by raulchen** [comment]
Can you also share how much faster compilation becomes?

**Comment by mehrdadn** [comment]
Unfortunately I think this by itself is not going to affect compilation speed noticeably. This is just a prerequisite for being to move out unnecessary headers, etc.

**Comment by mehrdadn** [comment]
But I may not merge this PR as-is, given people seem to be quite against this.

**Pull request closed without merging**